src/client/ua_client_subscriptions.c

```diff
@@ -111,61 +210,134 @@ UA_Client_Subscription_deleteInternal(UA_Client *client, UA_Client_Subscription
-    __UA_Client_Service(client,
-                        &request, &UA_TYPES[UA_TYPES_DELETESUBSCRIPTIONSREQUEST],
-                        &response, &UA_TYPES[UA_TYPES_DELETESUBSCRIPTIONSRESPONSE]);
-    if(response.responseHeader.serviceResult != UA_STATUSCODE_GOOD)
+static void __Subscriptions_delete_handler(UA_Client *client, void *data, UA_UInt32 requestId, void *r) {
+    UA_DeleteSubscriptionsResponse *response = (UA_DeleteSubscriptionsResponse *)r;
+    CustomCallback *cc = (CustomCallback *)data;
+    Subscriptions_DeleteData *delData = (Subscriptions_DeleteData *)cc->clientData;
+    UA_DeleteSubscriptionsRequest *request = delData->request;
+    UA_Client_Subscription **subs = delData->subs;
+
+    if (response->responseHeader.serviceResult != UA_STATUSCODE_GOOD)
         goto cleanup;
 
-    if(request.subscriptionIdsSize != response.resultsSize) {
-        response.responseHeader.serviceResult = UA_STATUSCODE_BADINTERNALERROR;
+    if (request->subscriptionIdsSize != response->resultsSize) {
+        response->responseHeader.serviceResult = UA_STATUSCODE_BADINTERNALERROR;
         goto cleanup;
     }
 
     /* Loop over the removed subscriptions and remove internally */
-    for(size_t i = 0; i < request.subscriptionIdsSize; i++) {
-        if(response.results[i] != UA_STATUSCODE_GOOD &&
-           response.results[i] != UA_STATUSCODE_BADSUBSCRIPTIONIDINVALID) {
+    for (size_t i = 0; i < request->subscriptionIdsSize; i++) {
+        if (response->results[i] != UA_STATUSCODE_GOOD &&
+            response->results[i] != UA_STATUSCODE_BADSUBSCRIPTIONIDINVALID) {
             /* Something was wrong, reinsert the subscription in the list */
             if (subs[i])
                 LIST_INSERT_HEAD(&client->subscriptions, subs[i], listEntry);
+            subs[i] = NULL;
             continue;
         }
 
         if(!subs[i]) {
-            UA_LOG_INFO(&client->config.logger, UA_LOGCATEGORY_CLIENT,
-                        "No internal representation of subscription %u",
-                        request.subscriptionIds[i]);
+            UA_LOG_INFO(&client->config.logger, UA_LOGCATEGORY_CLIENT, "No internal representation of subscription %u",
+                        delData->request->subscriptionIds[i]);
             continue;
         } else {
             LIST_INSERT_HEAD(&client->subscriptions, subs[i], listEntry);
         }
 
         UA_Client_Subscription_deleteInternal(client, subs[i]);
+        subs[i] = NULL;
     }
 
-    return response;
-
 cleanup:
-    for(size_t i = 0; i < request.subscriptionIdsSize; i++) {
+    for (size_t i = 0; i < request->subscriptionIdsSize; i++) {
         if (subs[i]) {
             LIST_INSERT_HEAD(&client->subscriptions, subs[i], listEntry);
         }
     }
+    if (cc->userCallback)
+        cc->userCallback(client, cc->userData, requestId, response);
+    CustomCallback_remove(cc, false);
+}
+
+UA_StatusCode UA_EXPORT UA_Client_Subscriptions_delete_async(UA_Client *client,
```

Comment:
UA_EXPORT is not needed in .c file.
Move method name to a new line.

Reply:
(same thing with clang format)
The existing functions did use UA_EXPORT too, should i remove UA_EXPORT them from my added new functions or from all in ua_client_subscriptions.c?

Reply:
UA_EXPORT is only for headers. Not for .c files.

Reply:
I have removed all UA_EXPORT occurrences in the ua_client_subscriptions.c file